lib/libc/newlib/libc-hooks.c

```diff
@@ -237,20 +238,33 @@ void _exit(int status)
 	}
 }
 
+/* TODO: Elide the mutex and do something clever with atomic operations
+ * to make _sbrk() thread-safe
+ */
+static LIBC_BSS SYS_MUTEX_DEFINE(heap_mutex);
```

Comment:
Maybe a spinlock (which becomes an irq_lock on uniprocessor systems) would be a better fit here?  This is a fast operation with bounded runtime.  There's no real benefit to trying to block a process on it and significant extra code involved in sucking in the mutex implementation.

Reply:
This code has to run in user mode, so a spinlock is not appropriate.

Reply:
Futex then? :)

Just seems like the (comparatively very heavyweight, especially when done via syscalls) blocking mechanism is exactly what we don't want when synchronizing a tiny bit of deterministic code in the guts of the allocator.

Reply:
@andyross that's what sys_mutex is.
Implementation isn't completely done to use atomic ops, but that is what it is for. See #15138

Reply:
I thought sys_mutex was the priority inheriting thing (which we don't use much from elsewhere in the kernel, will suck in non-trivial extra code for all newlib targets, and which is known to have SMP bugs)?

You win.

Reply:
> I thought sys_mutex was the priority inheriting thing 

It is. It is backed by a k_mutex on the kernel side. The design of sys mutex is to have the same semantics as FUTEX_LOCK_PI / FUTEX_UNLOCK_PI

> which is known to have SMP bugs

Which bugs are these?? We should label them high priority and fix them...

> will suck in non-trivial extra code for all newlib targets

What do you want here?
This code in its current state is broken.
Should I use a sys_sem instead?